<a href="https://colab.research.google.com/github/andrecarlos26/Projeto-Interdisciplinar-SI-2/blob/main/Pr%C3%A9_processamento_e_Tratamento.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:

from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
import pandas as pd
import statistics as sts
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, confusion_matrix
from sklearn.preprocessing import LabelEncoder


caminho = '/content/drive/My Drive/microdados_enem_2019/MICRODADOS_ENEM_2019_PE.csv'
dados_pe = pd.read_csv(caminho)

In [3]:
dados_pe.head()

NU_INSCRICAO  NU_ANO  CO_MUNICIPIO_RESIDENCIA  ... Q023  Q024 Q025
0  190001044305    2019                  2610004  ...    A     C    B
1  190001051159    2019                  2604007  ...    A     A    A
2  190001058610    2019                  2614105  ...    A     B    B
3  190001070444    2019                  2610509  ...    B     E    B
4  190001087987    2019                  2610707  ...    A     A    A

[5 rows x 136 columns]

# Tratamento de alunos faltosos e eliminados

In [4]:
# Encontrar index da feature
dados_pe.columns.get_loc('NU_NOTA_REDACAO')

110

In [5]:
# Remoção dos faltosos e eliminados do Primeiro dia (Redação, Linguagens e Humanas)
dados_pe.dropna(subset=['NU_NOTA_REDACAO'], inplace=True)

In [6]:
# Remoção dos faltosos e eliminados do Segundo Dia (Matemática e Natureza)
dados_pe.dropna(subset=['NU_NOTA_MT'], inplace=True)

In [7]:
# Tratamento de algumas colunas NaN de dados do participante que vão ser utilizadas no algoritmo de predição
moda = int(sts.mode(dados_pe['CO_MUNICIPIO_NASCIMENTO']))
dados_pe['CO_MUNICIPIO_NASCIMENTO'].fillna(moda, inplace=True)

In [8]:
moda = sts.mode(dados_pe['NO_MUNICIPIO_NASCIMENTO'])
dados_pe['NO_MUNICIPIO_NASCIMENTO'].fillna(moda, inplace=True)

In [9]:
moda = int(sts.mode(dados_pe['CO_UF_NASCIMENTO']))
dados_pe['CO_UF_NASCIMENTO'].fillna(moda, inplace=True)

In [10]:
moda = sts.mode(dados_pe['SG_UF_NASCIMENTO'])
dados_pe['SG_UF_NASCIMENTO'].fillna(moda, inplace=True)

In [11]:
dados_pe.shape

(202753, 136)

Após a remoção dos alunos faltosos de eliminados a quantidade de instâncias do dataset foi reduzida a 202753

#Criação das colunas NO_REGIAO_RESIDENCIA e NO_REGIAO_NASCIMENTO

In [12]:
dict_regioes_pe = { 
  'Araripina': 'Sertão',
  'Bodocó': 'Sertão',
  'Exu': 'Sertão',
  'Granito': 'Sertão',
  'Ipubi': 'Sertão',
  'Moreilândia': 'Sertão',
  'Ouricuri': 'Sertão',
  'Santa Cruz': 'Sertão',
  'Santa Filomena': 'Sertão',
  'Trindade': 'Sertão',
  'Cedro': 'Sertão',
  'Mirandiba': 'Sertão',
  'Parnamirim': 'Sertão',
  'Salgueiro': 'Sertão',
  'São José do Belmonte': 'Sertão',
  'Serrita': 'Sertão',
  'Verdejante': 'Sertão',
  'Afogados da Ingazeira': 'Sertão',
  'Brejinho': 'Sertão',
  'Calumbi': 'Sertão',
  'Carnaíba': 'Sertão',
  'Flores': 'Sertão',
  'Iguaracy': 'Sertão',
  'Ingazeira': 'Sertão',
  'Itapetim': 'Sertão',
  'Quixaba': 'Sertão',
  'Santa Cruz da Baixa Verde': 'Sertão',
  'Santa Terezinha': 'Sertão',
  'São José do Egito': 'Sertão',
  'Serra Talhada': 'Sertão',
  'Solidão': 'Sertão',
  'Tabira': 'Sertão',
  'Triunfo': 'Sertão',
  'Tuparetama': 'Sertão',
  'Arcoverde': 'Sertão',
  'Betânia': 'Sertão',
  'Custódia': 'Sertão',
  'Ibimirim': 'Sertão',
  'Inajá': 'Sertão',
  'Manari': 'Sertão',
  'Sertânia': 'Sertão',

  'Afrânio': 'São Francisco',
  'Cabrobó': 'São Francisco',
  'Dormentes': 'São Francisco',
  'Lagoa Grande': 'São Francisco',
  'Orocó': 'São Francisco',
  'Petrolina': 'São Francisco',
  'Santa Maria da Boa Vista': 'São Francisco',
  'Terra Nova': 'São Francisco',
  'Belém do São Francisco': 'São Francisco',
  'Carnaubeira da Penha': 'São Francisco',
  'Floresta': 'São Francisco',
  'Itacuruba': 'São Francisco',
  'Jatobá': 'São Francisco',
  'Petrolândia': 'São Francisco',
  'Tacaratu': 'São Francisco',

  'Águas Belas': 'Agreste',
  'Buíque': 'Agreste',
  'Itaíba': 'Agreste',
  'Pedra': 'Agreste',
  'Tupanatinga': 'Agreste',
  'Venturosa': 'Agreste',
  'Alagoinha': 'Agreste',
  'Belo Jardim': 'Agreste',
  'Bezerros': 'Agreste',
  'Brejo da Madre de Deus': 'Agreste',
  'Cachoeirinha': 'Agreste',
  'Capoeiras': 'Agreste',
  'Caruaru': 'Agreste',
  'Gravatá': 'Agreste',
  'Jataúba': 'Agreste',
  'Pesqueira': 'Agreste',
  'Poção': 'Agreste',
  'Riacho das Almas': 'Agreste',
  'Sanharó': 'Agreste',
  'São Bento do Una': 'Agreste',
  'São Caitano': 'Agreste',
  'Tacaimbó': 'Agreste',
  'Casinhas': 'Agreste',
  'Frei Miguelinho': 'Agreste',
  'Santa Cruz do Capibaribe': 'Agreste',
  'Santa Maria do Cambucá': 'Agreste',
  'Surubim': 'Agreste',
  'Taquaritinga do Norte': 'Agreste',
  'Toritama': 'Agreste',
  'Vertente do Lério': 'Agreste',
  'Vertentes': 'Agreste',
  'Bom Jardim': 'Agreste',
  'Cumaru': 'Agreste',
  'Feira Nova': 'Agreste',
  'João Alfredo': 'Agreste',
  'Limoeiro': 'Agreste',
  'Machados': 'Agreste',
  'Orobó': 'Agreste',
  'Passira': 'Agreste',
  'Salgadinho': 'Agreste',
  'São Vicente Férrer': 'Agreste',
  'Angelim': 'Agreste',
  'Bom Conselho': 'Agreste',
  'Brejão': 'Agreste',
  'Caetés': 'Agreste',
  'Calçado': 'Agreste',
  'Canhotinho': 'Agreste',
  'Correntes': 'Agreste',
  'Garanhuns': 'Agreste',
  'Iati': 'Agreste',
  'Jucati': 'Agreste',
  'Jupi': 'Agreste',
  'Jurema': 'Agreste',
  'Lagoa do Ouro': 'Agreste',
  'Lajedo': 'Agreste',
  'Palmeirina': 'Agreste',
  'Paranatama': 'Agreste',
  'Saloá': 'Agreste',
  'São João': 'Agreste',
  'Terezinha': 'Agreste',
  'Agrestina': 'Agreste',
  'Altinho': 'Agreste',
  'Barra de Guabiraba': 'Agreste',
  'Bonito': 'Agreste',
  'Camocim de São Félix': 'Agreste',
  'Cupira': 'Agreste',
  'Ibirajuba': 'Agreste',
  'Lagoa dos Gatos': 'Agreste',
  'Panelas': 'Agreste',
  'Sairé': 'Agreste',
  'São Joaquim do Monte': 'Agreste',

  'Aliança': 'Zona da Mata',
  'Buenos Aires': 'Zona da Mata',
  'Camutanga': 'Zona da Mata',
  'Carpina': 'Zona da Mata',
  'Condado': 'Zona da Mata',
  'Ferreiros': 'Zona da Mata',
  'Goiana': 'Zona da Mata',
  'Itambé': 'Zona da Mata',
  'Itaquitinga': 'Zona da Mata',
  'Lagoa de Itaenga': 'Zona da Mata',
  'Lagoa do Carro': 'Zona da Mata',
  'Macaparana': 'Zona da Mata',
  'Nazaré da Mata': 'Zona da Mata',
  'Paudalho': 'Zona da Mata',
  'Timbaúba': 'Zona da Mata',
  'Tracunhaém': 'Zona da Mata',
  'Vicência': 'Zona da Mata',
  'Chã de Alegria': 'Zona da Mata',
  'Chã Grande': 'Zona da Mata',
  'Glória do Goitá': 'Zona da Mata',
  'Pombos': 'Zona da Mata',
  'Vitória de Santo Antão': 'Zona da Mata',
  'Água Preta': 'Zona da Mata',
  'Amaraji': 'Zona da Mata',
  'Barreiros': 'Zona da Mata',
  'Belém de Maria': 'Zona da Mata',
  'Catende': 'Zona da Mata',
  'Cortês': 'Zona da Mata',
  'Escada': 'Zona da Mata',
  'Gameleira': 'Zona da Mata',
  'Jaqueira': 'Zona da Mata',
  'Joaquim Nabuco': 'Zona da Mata',
  'Maraial': 'Zona da Mata',
  'Palmares': 'Zona da Mata',
  'Primavera': 'Zona da Mata',
  'Quipapá': 'Zona da Mata',
  'Ribeirão': 'Zona da Mata',
  'Rio Formoso': 'Zona da Mata',
  'São Benedito do Sul': 'Zona da Mata',
  'São José da Coroa Grande': 'Zona da Mata',
  'Sirinhaém': 'Zona da Mata',
  'Tamandaré': 'Zona da Mata',
  'Xexéu': 'Zona da Mata',

  'Araçoiaba': 'Região Metropolitana',
  'Igarassu': 'Região Metropolitana',
  'Ilha de Itamaracá': 'Região Metropolitana',
  'Itapissuma': 'Região Metropolitana',
  'Abreu e Lima': 'Região Metropolitana',
  'Camaragibe': 'Região Metropolitana',
  'Jaboatão dos Guararapes': 'Região Metropolitana',
  'Moreno': 'Região Metropolitana',
  'Olinda': 'Região Metropolitana',
  'Paulista': 'Região Metropolitana',
  'Recife': 'Região Metropolitana',
  'São Lourenço da Mata': 'Região Metropolitana',
  'Cabo de Santo Agostinho': 'Região Metropolitana',
  'Ipojuca': 'Região Metropolitana',
  'Fernando de Noronha': 'Região Metropolitana',
}

In [13]:
regiao_residencia = []

for municipio in dados_pe['NO_MUNICIPIO_RESIDENCIA']:
  regiao_residencia.append(dict_regioes_pe[municipio])

dados_pe['NO_REGIAO_RESIDENCIA'] = regiao_residencia

In [14]:
regiao_nascimento = []

for municipio in dados_pe['NO_MUNICIPIO_NASCIMENTO']:
  try:
    regiao_nascimento.append(dict_regioes_pe[municipio])
  except KeyError:
    regiao_nascimento.append('Outro')

dados_pe['NO_REGIAO_NASCIMENTO'] = regiao_nascimento

# Criação das colunas NU_MEDIA e CATEGORIA_MEDIA

In [15]:
notas = dados_pe.iloc[:,  np.r_[91:95,110]].values

In [16]:
media = []

for i in notas:
  media.append(sts.mean(i))

In [17]:
dados_pe['NU_MEDIA'] = media

In [18]:
# obtenção dos valores dos quantis
quantil_25 = dados_pe['NU_MEDIA'].quantile(0.25)
quantil_50 = dados_pe['NU_MEDIA'].quantile(0.50)
quantil_75 = dados_pe['NU_MEDIA'].quantile(0.75)
print(quantil_25)
print(quantil_50)
print(quantil_75)

458.5
508.46
568.4200000000001


In [19]:
# Divisão para classificação binária
tipo_media = []
for i in media:
  if (i <= quantil_25):
    tipo_media.append('D')
  elif (i > quantil_25 and i <= quantil_50):
    tipo_media.append('C')
  elif (i > quantil_50 and i <= quantil_75):
    tipo_media.append('B')
  else:
    tipo_media.append('A')

In [20]:
dados_pe['TP_MEDIA'] = tipo_media

In [21]:
dados_pe.groupby('TP_MEDIA').size()

TP_MEDIA
A    50688
B    50678
C    50688
D    50699
dtype: int64

In [22]:
pd.set_option('max_columns', 138)
dados_pe.head()

NU_INSCRICAO  NU_ANO  CO_MUNICIPIO_RESIDENCIA NO_MUNICIPIO_RESIDENCIA  \
0   190001044305    2019                  2610004                Palmares   
5   190001095236    2019                  2604304                   Cedro   
7   190001098520    2019                  2615607                Trindade   
9   190001116648    2019                  2611101               Petrolina   
12  190001134974    2019                  2607307                   Ipubi   

    CO_UF_RESIDENCIA SG_UF_RESIDENCIA  NU_IDADE TP_SEXO  TP_ESTADO_CIVIL  \
0                 26               PE        21       F                0   
5                 26               PE        23       M                1   
7                 26               PE        19       M                1   
9                 26               PE        25       F                1   
12                26               PE        16       F                1   

    TP_COR_RACA  TP_NACIONALIDADE  CO_MUNICIPIO_NASCIMENTO  \
0             0                 1                2611606.0   
5             3                 1                2604304.0   
7             2                 1                2615607.0   
9             3                 1                2207801.0   
12            1                 1                2304202.0   

   NO_MUNICIPIO_NASCIMENTO  CO_UF_NASCIMENTO SG_UF_NASCIMENTO  \
0                   Recife              26.0               PE   
5                    Cedro              26.0               PE   
7                 Trindade              26.0               PE   
9               Paulistana              22.0               PI   
12                   Crato              23.0               CE   

    TP_ST_CONCLUSAO  TP_ANO_CONCLUIU  TP_ESCOLA  TP_ENSINO  IN_TREINEIRO  \
0                 1                4          1        1.0             0   
5                 1                5          1        1.0             0   
7                 1                2          1        1.0             0   
9                 1                6          1        NaN             0   
12                3                0          1        NaN             1   

    CO_ESCOLA  CO_MUNICIPIO_ESC NO_MUNICIPIO_ESC  CO_UF_ESC SG_UF_ESC  \
0         NaN               NaN              NaN        NaN       NaN   
5         NaN               NaN              NaN        NaN       NaN   
7         NaN               NaN              NaN        NaN       NaN   
9         NaN               NaN              NaN        NaN       NaN   
12        NaN               NaN              NaN        NaN       NaN   

    TP_DEPENDENCIA_ADM_ESC  TP_LOCALIZACAO_ESC  TP_SIT_FUNC_ESC  \
0                      NaN                 NaN              NaN   
5                      NaN                 NaN              NaN   
7                      NaN                 NaN              NaN   
9                      NaN                 NaN              NaN   
12                     NaN                 NaN              NaN   

    IN_BAIXA_VISAO  IN_CEGUEIRA  IN_SURDEZ  IN_DEFICIENCIA_AUDITIVA  \
0                0            0          0                        0   
5                0            0          0                        0   
7                0            0          0                        0   
9                0            0          0                        0   
12               0            0          0                        0   

    IN_SURDO_CEGUEIRA  IN_DEFICIENCIA_FISICA  IN_DEFICIENCIA_MENTAL  \
0                   0                      0                      0   
5                   0                      0                      0   
7                   0                      0                      0   
9                   0                      0                      0   
12                  0                      0                      0   

    IN_DEFICIT_ATENCAO  IN_DISLEXIA  IN_DISCALCULIA  IN_AUTISMO  \
0                    1            0               0           0   
5                    0           

# Verificando se existem instâncias fora do domínio

In [23]:
group_cor_raca = dados_pe.groupby(['TP_COR_RACA']).size()
group_cor_raca

TP_COR_RACA
0      4692
1     59917
2     23396
3    106420
4      5727
5      2601
dtype: int64

In [24]:
group_no_mun_res = dados_pe.groupby(['NO_REGIAO_RESIDENCIA']).size()
group_no_mun_res.sort_values(ascending=False)

NO_REGIAO_RESIDENCIA
Região Metropolitana    94189
Agreste                 45969
Zona da Mata            24459
Sertão                  19965
São Francisco           18171
dtype: int64

In [25]:
group_no_mun_nas = dados_pe.groupby(['NO_REGIAO_NASCIMENTO']).size()
group_no_mun_nas.sort_values(ascending=False)

NO_REGIAO_NASCIMENTO
Região Metropolitana    94074
Agreste                 39739
Zona da Mata            22311
Sertão                  18755
Outro                   15227
São Francisco           12647
dtype: int64

In [26]:
group_tp_ano_con = dados_pe.groupby(['TP_ANO_CONCLUIU']).size()
group_tp_ano_con

TP_ANO_CONCLUIU
0     93347
1     27233
2     17636
3     12314
4      8958
5      7003
6      5320
7      4233
8      3506
9      3012
10     2607
11     2047
12     1839
13    13698
dtype: int64

In [27]:
group_q001 = dados_pe.groupby(['Q001']).size()
group_q001

Q001
A    11722
B    46480
C    29615
D    21965
E    55944
F    12614
G     7829
H    16584
dtype: int64

In [28]:
group_q005 = dados_pe.groupby(['Q005']).size()
group_q005

Q005
1      3114
2     22003
3     50267
4     70136
5     35624
6     12776
7      5177
8      2130
9       746
10      425
11      155
12       97
13       42
14       24
15        9
16       10
17        3
18        6
19        1
20        8
dtype: int64

# Verificando se features que vão ser utilizadas possuem NaN's

In [29]:
pd.set_option('max_rows',300)
dados_pe.isnull().sum()

NU_INSCRICAO                        0
NU_ANO                              0
CO_MUNICIPIO_RESIDENCIA             0
NO_MUNICIPIO_RESIDENCIA             0
CO_UF_RESIDENCIA                    0
SG_UF_RESIDENCIA                    0
NU_IDADE                            0
TP_SEXO                             0
TP_ESTADO_CIVIL                     0
TP_COR_RACA                         0
TP_NACIONALIDADE                    0
CO_MUNICIPIO_NASCIMENTO             0
NO_MUNICIPIO_NASCIMENTO             0
CO_UF_NASCIMENTO                    0
SG_UF_NASCIMENTO                    0
TP_ST_CONCLUSAO                     0
TP_ANO_CONCLUIU                     0
TP_ESCOLA                           0
TP_ENSINO                       77189
IN_TREINEIRO                        0
CO_ESCOLA                      151981
CO_MUNICIPIO_ESC               151981
NO_MUNICIPIO_ESC               151981
CO_UF_ESC                      151981
SG_UF_ESC                      151981
TP_DEPENDENCIA_ADM_ESC         151981
TP_LOCALIZAC

# Extração de features selecionadas para predição

In [30]:
dados_pe.columns.get_loc('NO_REGIAO_RESIDENCIA')

136

In [31]:
dados_pe.columns.get_loc('NO_REGIAO_NASCIMENTO')

137

In [32]:
dados_pe.columns.get_loc('NU_IDADE')

6

In [33]:
dados_pe.columns.get_loc('TP_COR_RACA')

9

In [34]:
dados_pe.columns.get_loc('TP_ANO_CONCLUIU')

16

In [35]:
dados_pe.columns.get_loc('Q001')

111

In [36]:
dados_pe.columns.get_loc('Q006')

116

In [37]:
dados_pe.columns.get_loc('TP_MEDIA')

139

In [38]:
dados_pe.columns.get_loc('NU_MEDIA')

138

In [39]:
dados_selecionados = dados_pe.iloc[:,  np.r_[136, 137, 6, 9, 16, 111:117, 139, 138]]
dados_selecionados.head(10)

NO_REGIAO_RESIDENCIA  NO_REGIAO_NASCIMENTO  NU_IDADE  TP_COR_RACA  \
0          Zona da Mata  Região Metropolitana        21            0   
5                Sertão                Sertão        23            3   
7                Sertão                Sertão        19            2   
9         São Francisco                 Outro        25            3   
12               Sertão                 Outro        16            1   
13               Sertão                Sertão        17            1   
16               Sertão                Sertão        22            0   
17               Sertão                Sertão        19            3   
19               Sertão                Sertão        19            3   
20               Sertão                Sertão        17            1   

    TP_ANO_CONCLUIU Q001 Q002 Q003 Q004  Q005 Q006 TP_MEDIA  NU_MEDIA  
0                 4    E    H    F    F     6    E        A    701.82  
5                 5    E    C    A    A     4    B        D    437.34  
7                 2    E    C    B    A     6    C        A    655.60  
9                 6    H    B    A    A     4    B        B    526.34  
12                0    C    F    D    D     5    D        A    651.76  
13                0    B    B    F    A     3    B        A    634.92  
16                6    A    D    A    A     4    B        B    539.90  
17                3    B    E    A    B     4    B        B    551.50  
19                2    F    F    D    D     5    E        A    587.12  
20                0    B    E    A    A     3    B        C    494.44

In [40]:
dados_selecionados.index = range(len(dados_selecionados.index))
dados_selecionados

NO_REGIAO_RESIDENCIA  NO_REGIAO_NASCIMENTO  NU_IDADE  TP_COR_RACA  \
0               Zona da Mata  Região Metropolitana        21            0   
1                     Sertão                Sertão        23            3   
2                     Sertão                Sertão        19            2   
3              São Francisco                 Outro        25            3   
4                     Sertão                 Outro        16            1   
...                      ...                   ...       ...          ...   
202748               Agreste               Agreste        28            3   
202749  Região Metropolitana  Região Metropolitana        19            3   
202750                Sertão                Sertão        21            1   
202751  Região Metropolitana  Região Metropolitana        44            1   
202752  Região Metropolitana  Região Metropolitana        19            3   

        TP_ANO_CONCLUIU Q001 Q002 Q003 Q004  Q005 Q006 TP_MEDIA  NU_MEDIA  
0                     4    E    H    F    F     6    E        A    701.82  
1                     5    E    C    A    A     4    B        D    437.34  
2                     2    E    C    B    A     6    C        A    655.60  
3                     6    H    B    A    A     4    B        B    526.34  
4                     0    C    F    D    D     5    D        A    651.76  
...                 ...  ...  ...  ...  ...   ...  ...      ...       ...  
202748               10    B    B    A    A     5    A        D    448.04  
202749                3    E    E    B    B     2    C        B    545.56  
202750                5    C    F    A    D     8    E        B    562.10  
202751                8    A    A    B    B     3    B        C    486.16  
202752                3    A    B    C    B     5    E        A    597.22  

[202753 rows x 13 columns]

# Exportação do novo dataframe csv

In [41]:
dados_selecionados.to_csv('/content/drive/My Drive/microdados_enem_2019/DADOS_TRATADOS_QUANTIL.csv', index=False) 